src/build/__main__.py

```diff
@@ -126,6 +119,56 @@ def build_package(srcdir, outdir, distributions, config_settings=None, isolation
         _error(str(e))
 
 
+def build_package(srcdir, outdir, distributions, config_settings=None, isolation=True, skip_dependency_check=False):
+    # type: (str, str, List[str], Optional[ConfigSettingsType], bool, bool) -> None
+    """
+    Run the build process.
+
+    :param srcdir: Source directory
+    :param outdir: Output directory
+    :param distribution: Distribution to build (sdist or wheel)
+    :param config_settings: Configuration settings to be passed to the backend
+    :param isolation: Isolate the build in a separate environment
+    :param skip_dependency_check: Do not perform the dependency check
+    """
+    with _handle_build_error():
+        builder = ProjectBuilder(srcdir)
+        for distribution in distributions:
+            _build(isolation, builder, outdir, distribution, config_settings, skip_dependency_check)
+
+
+def build_package_via_sdist(srcdir, outdir, distributions, config_settings=None, isolation=True, skip_dependency_check=False):
+    # type: (str, str, List[str], Optional[ConfigSettingsType], bool, bool) -> None
+    """
+    Build a sdist and then the specified distributions from it.
+
+    :param srcdir: Source directory
+    :param outdir: Output directory
+    :param distribution: Distribution to build (only wheel)
+    :param config_settings: Configuration settings to be passed to the backend
+    :param isolation: Isolate the build in a separate environment
+    :param skip_dependency_check: Do not perform the dependency check
+    """
+    if 'sdist' in distributions:
+        raise ValueError('Only binary distributions are allowed but sdist was specified')
+
+    with _handle_build_error():
+        builder = ProjectBuilder(srcdir)
+        sdist = _build(isolation, builder, outdir, 'sdist', config_settings, skip_dependency_check)
+
+        # extract sdist
+        sdist_name = os.path.basename(sdist)
+        sdist_out = tempfile.mkdtemp(dir=outdir, prefix='build-via-sdist-')
```

Comment:
Build artefacts should not be placed in `outdir`, suggesting removing the `dir` argument.

Reply:
Why not? See my comment below, I don't want to put it in a temporary path as that is subject to be removed, and leaving stuff behind in such paths is not very hygienic.

Reply:
Conceptually, for much the same reasons that `bdist_wheel` makes a copy of the source tree in `build` and not `dist` - the output directory's there for the final build artefacts, not the bits and bobs that go into making them.  Besides being outside of the norm, it might also trip other programs up that don't do any filtering, especially considering we don't (yet) have an answer to #198. For example, this is what happens when you `twine upload dist/*`, where `dist` contains a folder:

```
$ twine upload dist/*
Uploading distributions to https://upload.pypi.org/legacy/
InvalidDistribution: Unknown distribution format: 'foo'
```

Reply:
But build artifacts are only left there if the build fails. I agree that we should not be keeping ephemeral build artifacts there on successful builds, but IMO it would be reasonable to put them there on failed builds.
A successful run should leave you with only the final artifacts, an unsuccessful one will leave you with the information needed to debug the issue.

Reply:
Also, I'd like to point out this behavior is common in many other tools.

Reply:
Fair enough, let's do it that way.

Reply:
This will break a lot of assumptions - it's the cause for #320, and likely will interfere with SCM versioning, etc. These need to be placed in an isolated output directory, not in source. Most other tools have a "--dirty" command that keeps from removing the temporary build directory if needed for debugging.

If you want a failed build to produce something detectable, maybe copy the contents of the temp dir to `<outdir>` on failure? Though leaving stuff around by default doesn't seem like a good idea - most users won't know that they need to clean up and the output dir could grow.

Reply:
Right, I failed to properly consider those use-cases. Let's do the build in an isolated directory.